GitUI/Editor/BlameAuthorMargin.cs

```diff
@@ -0,0 +1,119 @@
+        public int RecencyIndex { get; set; }
+        public Color RecencyColor { get; set; }
+    }
+
+    /// <summary>
+    /// This class display avatars in the gutter in a blame control.
+    /// </summary>
+    public class BlameAuthorMargin : AbstractMargin
+    {
+        private static readonly int RecencyIndexWidth = Convert.ToInt32(4 * DpiUtil.ScaleX);
+        private List<Image> _avatars;
+        private readonly int _lineHeight;
+        private readonly Color _backgroundColor;
+        private List<GitBlameDisplay> _blameLines;
+        private readonly Dictionary<int, SolidBrush> _brushs = new Dictionary<int, SolidBrush>();
+        private bool _isVisible = true;
+
+        public BlameAuthorMargin(TextArea textArea) : base(textArea)
+        {
+            _lineHeight = GetFontHeight(textArea.Font);
+            _backgroundColor = ((SolidBrush)SystemBrushes.Window).Color;
+            Width = _lineHeight + RecencyIndexWidth + DpiUtil.Scale(2);
+        }
+
+        private static int GetFontHeight(Font font)
+        {
+            var max = Math.Max(
+                TextRenderer.MeasureText("_", font).Height,
+                (int)Math.Ceiling(font.GetHeight()));
+
+            return max + 1;
+        }
+
+        public void Initialize(List<GitBlameDisplay> blameLines)
+        {
+            _blameLines = blameLines;
+            _avatars = blameLines.Select(a => ThreadHelper.JoinableTaskFactory.Run(async () => await a.Avatar)).ToList();
+
+            // Update the resolution otherwise the image is not drawn at the good size :(
+            foreach (var avatar in _avatars)
+            {
+                if (avatar is Bitmap bitmapAvatar)
+                {
+                    bitmapAvatar.SetResolution(DpiUtil.DpiX, DpiUtil.DpiY);
+                }
+            }
+
+            // Build brushes
+            foreach (var blameLine in blameLines)
+            {
+                if (!_brushs.ContainsKey(blameLine.RecencyIndex))
+                {
+                    _brushs.Add(blameLine.RecencyIndex, new SolidBrush(blameLine.RecencyColor));
+                }
+            }
+        }
+
+        public override int Width { get; }
+
+        public void SetVisiblity(bool isVisible)
+        {
+            _isVisible = isVisible;
+        }
+
+        public override bool IsVisible => _isVisible;
+
+        public override void Paint(Graphics g, Rectangle rect)
+        {
+            if (rect.Width <= 0 || rect.Height <= 0)
+            {
+                return;
+            }
+
+            g.Clear(_backgroundColor);
+
+            if (_avatars == null || _avatars.Count == 0)
+            {
+                return;
+            }
+
+            var verticalOffset = textArea.VirtualTop.Y;
+            var lineStart = verticalOffset / _lineHeight;
+            var negativeOffset = (lineStart * _lineHeight) - verticalOffset;
+            var lineCount = (int)rect.Height / _lineHeight;
```

Comment:
Correct me if I am wrong, I believe `lineCount` should be rather
```csharp
var lineCount = (int)Math.Ceiling((rect.Height - negativeOffset) / _lineHeight);
```
`- negativeOffset` because the first line may start above 0 due to `negativeOffset`
`(int)Math.Ceiling()` because we should also paint the bottom partially visible row